APISyncExternalModule.php

```diff
@@ -898,25 +905,291 @@ function cancelSync(){
+						$instance['redcap_repeat_instrument'] = $local_form_name;
+					}
+					
+					// [$form . '_complete'] components
+					foreach($form_names as $other_form_name) {
+						if (isset($instance[$other_form_name . '_complete'])) {
+							$instance[$local_form_name . '_complete'] = $instance[$other_form_name . '_complete'];
+							unset($instance[$other_form_name . '_complete']);
+						}
+					}
+				}
+			}
+		}
+	}
+	
+	public function translateEventNames(&$data, &$project) {
+		$proj_prefix = $this->getProjectTypePrefix($project);
+		$translations = $project[$proj_prefix . 'event-translations'];
+		if (gettype($translations) != 'array') {
+			return;
+		}
+		// translate [redcap_event_name] fields where applicable
+		foreach ($data as &$instance) {
+			foreach ($translations as $event_names) {
+				// [redcap_event_name] component
+				$local_event_name = $event_names[0];
+				$export_event_name = $event_names[1];
+				$event_name_pieces = explode('_arm_', $instance['redcap_event_name']);
+				$event_name = $event_name_pieces[0];
+				$event_arm_number = $event_name_pieces[1];
+				if ($proj_prefix == 'export-') {
+					// export logic
+					if ($event_name == $local_event_name) {
+						$instance['redcap_event_name'] = $export_event_name . "_arm_" . $event_arm_number;
+						break;
+					}
+				} else {
+					// import logic
+					if (in_array($event_name, $event_names, true)) {
+						$instance['redcap_event_name'] = $local_event_name . "_arm_" . $event_arm_number;
+						break;
+					}
+				}
+			}
+		}
+	}
+	
+	public function buildTranslations(&$project) {
+		// function will only build translations if json present in $project['form-translations'/'event-translations']
+		if ($this->translationsAreBuilt($project)) {
+			return;
+		}
+		$proj_key_prefix = $this->getProjectTypePrefix($project);
+		
+		foreach (['form', 'event'] as $type) {
+			$setting = &$project[$proj_key_prefix . "$type-translations"];
+			$setting = json_decode($setting, true);
+			if ($setting) {
+				foreach($setting as $i => $row) {
+					foreach($row as $j => $name) {
+						$func_name = "format" . ucfirst($type) . "Name";
+						$setting[$i][$j] = $this->$func_name($name);
+					}
+				}
+			}
+		}
+	}
 
-		$result = $this->query("
-			select field_name
-			from redcap_metadata
-			where project_id = $pid
-			order by field_order
-			limit 1
-		");
+	private function translationsAreBuilt($project) {
+		$translation_settings = [
+			'form-translations',
+			'event-translations',
+			'export-form-translations',
+			'export-event-translations'
+		];
+		foreach($translation_settings as $name) {
+			if (gettype($project[$name]) == 'array') {
+				return true;
+			}
+		}
+	}
 
-		$row = $result->fetch_assoc();
+	private function getProjectTypePrefix(&$project) {
+		foreach ($project as $setting_name => $setting_value) {
+			if (strpos($setting_name, 'export-') !== false) {
+				return 'export-';
+			}
+		}
```

Comment:
Interesting...  I'm not sure I would have thought to do this, but it seems effective.

Reply:
Yeah. This is not my favorite thing but it does work.